### PR TITLE
test: add EXPECT_EMPTY_RESULT

### PR DESCRIPTION
**Pull request by axic**
Merge after #255 and #259.

### CODE REVIEW

File: test/unittests/execute_test.cpp [hunk 1]
```diff
@@ -13,9 +13,7 @@ TEST(execute, end)
     (func)
     */
     const auto wasm = from_hex("0061736d01000000010401600000030201000a040102000b");
-    const auto [trap, ret] = execute(parse(wasm), 0, {});
-    ASSERT_FALSE(trap);
-    EXPECT_EQ(ret.size(), 0);
+    EXPECT_EMPTY_RESULT(execute(parse(wasm), 0, {}));
```
Comment (axic) [inline test/unittests/execute_test.cpp:16]: How about merging `EXPECT_*_RESULT` to expect an array and deduce the size/values from it?

```
EXPECT_RESULT(execute(parse(wasm), 0, {}), {})
EXPECT_RESULT(execute(parse(wasm), 0, {}), {1})
EXPECT_RESULT(execute(parse(wasm), 0, {}), {1, 2})
```

@chfast @gumb0 what od you think?

File: test/utils/asserts.hpp [hunk 2]
```diff
@@ -11,6 +11,14 @@
         EXPECT_EQ(r.stack[0], (expected)); \
     } while (false)
 
+#define EXPECT_EMPTY_RESULT(result)   \
+    do                                \
+    {                                 \
+        const auto r = (result);      \
+        ASSERT_FALSE(r.trapped);      \
```
Comment (gumb0) [inline test/utils/asserts.hpp:18]: It's called `EXPECT_` but in fact does `ASSERT`.
I think I would change both `ASSERT`s to `EXPECT`s

Comment (axic) [reply test/utils/asserts.hpp:18]: Yeah assert vs expect is plagued all over the test suite. if we want to be fully consistent then trapped should still be assert, but the size can be changed to expect. (See `EXPECT_RESULT`).

File: test/unittests/execute_test.cpp (hunk 1 above)
Comment (chfast) [reply test/unittests/execute_test.cpp:16]: Macros don't go along with `{}`: `EXPECT_RESULT(execute(parse(wasm), 0, {}), {1, 2})` will be parsing error. You need `EXPECT_RESULT(execute(parse(wasm), 0, {}), ({1, 2}))`.
I also don't expect returning all remaining stack will stay with us for long. You may try to get it work with `std::optional<uint64_t>`.

**Comment by chfast** [comment]
If we are here, I think it's better to migrate/use `EXPECT_THAT()` from gmock. They work if you include `#include <gmock/gmock-matchers.h>`.

- https://github.com/google/googletest/blob/master/googlemock/docs/cook_book.md#using-matchers
- https://github.com/google/googletest/blob/master/googlemock/docs/cook_book.md#writing-new-matchers-quickly-newmatchers

Then we can implement `Result()` matcher instead of fighting with macros and use it as
```
EXPECT_THAT(execute(instance, 0, {}), Result());
EXPECT_THAT(execute(instance, 1, {}), Result(123));
```

**Comment by axic** [comment]
> If we are here, I think it's better to migrate/use EXPECT_THAT() from gmock. 

Sounds good to me.

File: test/unittests/execute_test.cpp [hunk 3]
```diff
@@ -1,21 +1,37 @@
...
 #include "limits.hpp"
 #include "parser.hpp"
+#include <gmock/gmock.h>
 #include <gtest/gtest.h>
 #include <test/utils/asserts.hpp>
 #include <test/utils/hex.hpp>
 
 using namespace fizzy;
 
+MATCHER(Trapped, "")
```
Comment (axic) [inline test/unittests/execute_test.cpp:11]: Are we adding these to `asserts.hpp`?

Comment (chfast) [reply test/unittests/execute_test.cpp:11]: Yes.

Comment (chfast) [inline test/unittests/execute_test.cpp:11]: I expect that `""` is used in failure messages. Please check how it looks add descriptions.

Comment (axic) [reply test/unittests/execute_test.cpp:11]: Yes, the manual said that. Will need to come up with something.

Comment (axic) [reply test/unittests/execute_test.cpp:11]:
```
/Users/alex/Projects/fizzy/test/unittests/execute_test.cpp:34: Failure
Value of: execute(parse(wasm), 0, {})
Expected: trapped
  Actual: 32-byte object <00-CA 7C-0F 01-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00> (of type fizzy::execution_result)
```

It seems matchers are just lowercased. However need to have some kind of formatter for `execution_result`.

File: test/unittests/execute_test.cpp [hunk 4]
```diff
@@ -1,21 +1,37 @@
 #include "execute.hpp"
 #include "limits.hpp"
 #include "parser.hpp"
+#include <gmock/gmock.h>
 #include <gtest/gtest.h>
```
Comment (gumb0) [inline test/unittests/execute_test.cpp:5]: Is including just `gmock.h` enough?

File: test/unittests/execute_test.cpp (hunk 3 above)
Comment (gumb0) [inline test/unittests/execute_test.cpp:11]: following other matchers in gmock
```suggestion
MATCHER(Traps, "traps")
```

File: test/unittests/execute_test.cpp [hunk 5]
```diff
@@ -1,21 +1,37 @@
...
+}
+MATCHER(Result, "")
+{
+    return arg.trapped == false && arg.stack.size() == 0;
+}
+MATCHER_P(Result, value, "")
+{
+    return arg.trapped == false && arg.stack.size() == 1 && arg.stack[0] == uint64_t(value);
+}
+MATCHER_P2(Result, value1, value2, "")
```
Comment (gumb0) [inline test/unittests/execute_test.cpp:23]: Where do we need this?

Comment (axic) [reply test/unittests/execute_test.cpp:23]: There were some tests doing it, but couldn't find them yesterday anymore.

**Pull request closed without merging**